prometheus_juju_backup_all_exporter/exporter.py

```diff
@@ -0,0 +1,75 @@
+import socket
+import threading
+from logging import getLogger
+from socketserver import ThreadingMixIn
+from wsgiref.simple_server import WSGIRequestHandler, WSGIServer, make_server
+
+from prometheus_client import make_wsgi_app
+from prometheus_client.core import REGISTRY
+
+logger = getLogger(__name__)
+
+
+def _get_best_family(address, port):
+    """Automatically select address family depending on address."""
+    # Copied from
+    # https://github.com/prometheus/client_python/blob/master/prometheus_client/exposition.py#L152
+    infos = socket.getaddrinfo(address, port)
+    family, _, _, _, sockaddr = next(iter(infos))
+    return family, sockaddr[0]
+
+
+class ThreadedWSGIServer(ThreadingMixIn, WSGIServer):
+    """A WSGI server that handle requests in a separate thread."""
+
+
+class SlientRequestHandler(WSGIRequestHandler):
+    """A Slient Request handler."""
+
+    def log_message(self, format, *args):
+        """Log nothing."""
+        pass
+
+
+class Exporter(object):
+    """The exporter class."""
+
+    def __init__(self, port, addr="0.0.0.0"):
+        """Initialize the exporter class.
+
+        Args:
+            port: Start the exporter at this port.
+        """
+        self.port = int(port)
+        self.addr = str(addr)
+        self.app = make_wsgi_app()
+
+        class ThreadedExporterServer(ThreadedWSGIServer):
```

Comment:
What's the benefit of doing all this manual setup with `ThreadedExporterServer`, `make_wsgi_app` and `make_server` as opposed to using simple [`start_http_server(port, addr)` from prometheus_client](https://github.com/prometheus/client_python#three-step-demo) ?

Reply:
The only problem here is that the built-in `start_http_server` is a daemon thread, it will exit once the program exits (in this case, `snap start <exporter>` will immediately exit the http server). We can easily fix it by adding a `while True:` loop, but Eric and I think that's probably no a nice solution.

Reply:
Aren't you gonna need another thread (presumably the "main") alive anyway to make updates to the exported data?

Reply:
No, I think it isn't the best practice to have your own scheduler for updating exported data. See: https://prometheus.io/docs/instrumenting/writing_exporters/#scheduling

Reply:
Oh, interesting. Thanks for info. I wouldn't have guessed that his is preferred workflow. In that case, I'd suggest two minor tweaks.

1. re-assignment of the `self.addr` seems superfluous. You are setting it on line 44 and then immediately re-setting it on line 50.

2. And I feel like this is bigger issue. The "copy" class `ThreadedExporterServer` scoped within the `__init__` is confusing imo. Is it done so that you wouldn't change `address_family` class variable in `ThreadedWSGIServer`? (could you also please explain the purpose of `address_family`? I tried to look it up in [WSGIServer docs](https://docs.python.org/3/library/wsgiref.html#wsgiref.simple_server.WSGIServer) but I didn't find anything.)

Reply:
For 1, thanks for spotting this, I didn't realized that ;) 
For 2, I didn't invented this function. I basically "copy" the code from the `prometheus_client` library. https://github.com/prometheus/client_python/blob/master/prometheus_client/exposition.py#L152. There seems to be some issue (I did not dig into this issue).

Reply:
re 2:
Fair enough. Though in your case I don't think you need to worry with the "copy" class. You can modify `ThreadedWSGIServer` directly as you won't be creating multiple instances with different address families anyway, right? It's up to you but I think it'd clean up the code a litle.

Reply:
I've created a helper function to generate the class. This should clean up the `__init__` method.